jax/interpreters/xla.py

```diff
@@ -318,6 +369,37 @@ class DeviceValue(object):
   def __init__(self, device_buffer):
     self.device_buffer = device_buffer
 
+
+class DeviceTuple(DeviceValue):
+  __slots__ = ["elt_xla_shapes", "elt_handlers"]
+  def __init__(self, device_buffer):
+    self.device_buffer = device_buffer
+    self.elt_xla_shapes = device_buffer.shape().tuple_shapes()
+    self.elt_handlers = list(map(_tuple_elt_handler, self.elt_xla_shapes))
+
+  def __iter__(self):
+    bufs = self.device_buffer.destructure()
+    elts = [handler(buf) for handler, buf in zip(self.elt_handlers, bufs)]
+    return iter(elts)
+
+  def __len__(self):
+    return len(self.elt_xla_shapes)
+
+def _tuple_elt_handler(xla_shape):
+  if xla_shape.is_tuple():
+    return DeviceTuple
+  else:
+    result_shape = xla_shape_to_result_shape(xla_shape)
+    return lambda buf: DeviceArray(buf, *result_shape)
+
+def abstractify_device_tuple(tup):
+  return AbstractTuple(map(aval_from_xla_shape, tup.elt_xla_shapes))
+
+core.pytype_aval_mappings[DeviceTuple] = AbstractTuple
+pytype_aval_mappings[DeviceTuple] = abstractify_device_tuple
+canonicalize_dtype_handlers[DeviceTuple] = identity
```

Comment:
How is canonicalization defined in the presence of tuples?
e.g., (float64, float64) should presumably become (float32, float32) in non-x64 mode?

Shouldn't this recurse?

Reply:
The theory here is that if something's already on the device then it's already canonicalized. We use analogous logic in setting the canonicalization handler for `DeviceArray` to be `identity` as well.

This relies on the fact that the only way to create a DeviceArray/DeviceTuple is via a path that does canonicalization. At the user level (in api.py), that's handled by the fact that (1) api.device_put is just `jit(lambda x: x)` and (2) jitted functions have their arguments canonicalized. (I believe the only callers of xla.device_put are the argument handlers in xla.py and in pxla.py, and both of those canonicalize before putting.)

Does that make sense?

Reply:
That makes sense but deserves a comment.

Reply:
Added a comment!